src/storage/storage_structure/node_column.cpp

```diff
@@ -0,0 +1,468 @@
+    nodeGroupsDataFH->releaseWALPageIdxLock(walPageInfo.originalPageIdx);
+}
+
+void NullNodeColumn::writeInternal(
+    offset_t nodeOffset, ValueVector* vectorToWriteFrom, uint32_t posInVectorToWriteFrom) {
+    writeSingleValue(nodeOffset, vectorToWriteFrom, posInVectorToWriteFrom);
+}
+
+SerialNodeColumn::SerialNodeColumn(const catalog::MetaDiskArrayHeaderInfo& metaDAHeaderInfo,
+    BMFileHandle* nodeGroupsDataFH, BMFileHandle* nodeGroupsMetaFH, BufferManager* bufferManager,
+    WAL* wal)
+    : NodeColumn{LogicalType(LogicalTypeID::SERIAL), metaDAHeaderInfo, nodeGroupsDataFH,
+          nodeGroupsMetaFH, bufferManager, wal, false} {}
+
+void SerialNodeColumn::scan(
+    Transaction* transaction, ValueVector* nodeIDVector, ValueVector* resultVector) {
+    // Serial column cannot contain null values.
+    for (auto i = 0ul; i < nodeIDVector->state->selVector->selectedSize; i++) {
+        auto pos = nodeIDVector->state->selVector->selectedPositions[i];
+        auto offset = nodeIDVector->readNodeOffset(pos);
+        resultVector->setValue<offset_t>(pos, offset);
+    }
+}
+
+void SerialNodeColumn::lookup(
+    Transaction* transaction, ValueVector* nodeIDVector, ValueVector* resultVector) {
+    if (nodeIDVector->state->isFlat()) {
+        // Serial column cannot contain null values.
+        auto pos = nodeIDVector->state->selVector->selectedPositions[0];
+        auto offset = nodeIDVector->readNodeOffset(pos);
+        resultVector->setValue<offset_t>(pos, offset);
+    } else {
+        // Serial column cannot contain null values.
+        for (auto i = 0ul; i < nodeIDVector->state->selVector->selectedSize; i++) {
+            auto pos = nodeIDVector->state->selVector->selectedPositions[i];
+            auto offset = nodeIDVector->readNodeOffset(pos);
+            resultVector->setValue<offset_t>(pos, offset);
+        }
+    }
+}
+
+page_idx_t SerialNodeColumn::appendColumnChunk(
+    ColumnChunk* columnChunk, page_idx_t startPageIdx, uint64_t nodeGroupIdx) {
+    // DO NOTHING.
+    return 0;
+}
+
+std::unique_ptr<NodeColumn> NodeColumnFactory::createNodeColumn(const LogicalType& dataType,
+    const catalog::MetaDiskArrayHeaderInfo& metaDAHeaderInfo, BMFileHandle* nodeGroupsDataFH,
+    BMFileHandle* nodeGroupsMetaFH, BufferManager* bufferManager, WAL* wal) {
+    switch (dataType.getLogicalTypeID()) {
+    case LogicalTypeID::BOOL:
+    case LogicalTypeID::INT64:
+    case LogicalTypeID::INT32:
+    case LogicalTypeID::INT16:
+    case LogicalTypeID::DOUBLE:
+    case LogicalTypeID::FLOAT:
+    case LogicalTypeID::DATE:
+    case LogicalTypeID::TIMESTAMP:
+    case LogicalTypeID::INTERVAL:
+    case LogicalTypeID::INTERNAL_ID:
+    case LogicalTypeID::FIXED_LIST: {
+        return std::make_unique<NodeColumn>(dataType, metaDAHeaderInfo, nodeGroupsDataFH,
+            nodeGroupsMetaFH, bufferManager, wal, true);
+    }
+        // TODO: Add a special case for FIXED_LIST, which should read without assuming 2^n per val.
```

Comment:
Again let's keep track of the TODOs. Also the wrapping looks off.

Reply:
This TODO is not relevant. Removed it.